init FIPS documentation (#11412)
opensearch-project/documentation-website: _security/configuration/fips.md

---
layout: default
title: Getting started with OpenSearch FIPS
nav_order: 70
---

# Getting started with OpenSearch FIPS

The Federal Information Processing Standard (FIPS) 140-2 is a U.S. government standard that defines security requirements for cryptographic modules. When running OpenSearch in a FIPS-compliant environment, you need to configure the system to use FIPS-validated cryptographic providers.

To achieve FIPS compliance, OpenSearch requires:

- FIPS-validated cryptographic providers for all cryptographic operations (Bouncy Castle FIPS is included with OpenSearch)
- JVM configured to use these FIPS-validated providers
- FIPS-compliant key stores and trust stores in BCFKS or PKCS11 format
- Strong passwords meeting FIPS minimum requirements (112 bits / approximately 14 characters)

## FIPS demo installer

> By default, the JVM uses the `cacerts` trust store (typically in PKCS12 format) for SSL/TLS connections, which contains trusted certificate authority (CA) certificates. However, the standard PKCS12 format is not FIPS-compliant.

Author: Since we actively prohibit any truststore other than PKCS11 and BCFKS, it’s fair to point out that PKCS12 isn’t supported.
https://github.com/opensearch-project/OpenSearch/blob/0449ce85aff3c3f307bec38b58b8928a8bd6d995/server/src/main/java/org/opensearch/bootstrap/FipsTrustStoreValidator.java#L86
@terryquigleysas Are you okey to mark it as resolved?

By default, the JVM uses the `cacerts` trust store (typically in PKCS12 format) for SSL/TLS connections, which contains trusted certificate authority (CA) certificates. However, the standard PKCS12 format is not FIPS-compliant.

OpenSearch includes a FIPS demo installer CLI tool that simplifies the trust store configuration process. The tool provides an automated way to set up a FIPS-compliant trust store by converting the JVM's default trust store to BCFKS format. The project source is available in distribution/tools/fips-demo-installer-cli.

This tool is designed for demo and development purposes. Before deploying to production, carefully review all generated configuration and replace demo settings with production-appropriate values.
{: .warning}

### Prerequisites

Before running the FIPS demo installer, ensure the following:

- OpenSearch is installed and the installation directory is accessible.
- You have write permissions to the OpenSearch configuration directory.
- The `jvm.options` file exists in the configuration directory.

### Available commands

The FIPS demo installer provides the following commands:

| Command | Description |
|---------|-------------|

> | `generated` | Generate a new BCFKS trust store from the JVM default trust store |

Author: actually yes :)

Because FIPS mode is already active, the `opensearch-keystore passwd` command will not accept the existing weak password. The workaround is to recreate the keystore:

```bash
# List existing secrets for backup (if needed)
./bin/opensearch-keystore list

# Create a new keystore with a FIPS-compliant password (at least 14 characters)
./bin/opensearch-keystore create --password

# Re-add any secrets that were stored in the old keystore (if needed)
./bin/opensearch-keystore add <setting-name>
```
{% include copy.html %}

Ensure your new password is at least 14 characters long and includes a mix of uppercase, lowercase, numbers, and special characters. For security best practices, consider using a password manager to generate and store complex passwords.
{: .note}

## Next steps

After configuring FIPS mode for OpenSearch:

> - Review the [Security configuration]({{site.url}}{{site.baseurl}}/security/configuration/index/) guide for additional security settings.

Author: good idea!

---
layout: default
title: OpenSearch FIPS configuration
parent: Configuration
nav_order: 20
---

# OpenSearch FIPS configuration

The [Federal Information Processing Standard (FIPS) 140-3](https://csrc.nist.gov/pubs/fips/140-3/final) is a U.S. government standard that defines security requirements for cryptographic modules. When running OpenSearch in a FIPS-compliant environment, you must configure the system to use FIPS-validated cryptographic providers.

To achieve FIPS compliance, OpenSearch requires:

- FIPS-validated cryptographic providers for all cryptographic operations (Bouncy Castle FIPS is included in OpenSearch)

> - JVM configured to use these FIPS-validated providers

Reviewer: ```suggestion
- A JVM configured to use these FIPS-validated providers.
```

> - JVM using a Java version for which [BC-FJA](https://www.bouncycastle.org/download/bouncy-castle-java-fips/) is certified (for example, Java 11, 17, or 21 for OpenSearch v3.2.0+ with BC-FJA v2.1.0)

Reviewer: ```suggestion
- A JVM using a Java version for which [BC-FJA](https://www.bouncycastle.org/download/bouncy-castle-java-fips/) is certified (for example, Java 11, 17, or 21 for OpenSearch v3.2.0+ with BC-FJA v2.1.0).
```

> - FIPS-compliant keystores and truststores in BCFKS or PKCS11 format

Reviewer: ```suggestion
- FIPS-compliant keystores and truststores in BCFKS or PKCS11 format.
```

> - Strong passwords meeting FIPS minimum requirements (112 bits or approximately 14 characters)

Reviewer: ```suggestion
- Strong passwords meeting FIPS minimum requirements (112 bits or approximately 14 characters).
```

OpenSearch includes a FIPS demo installer CLI tool that simplifies the truststore configuration process. The tool provides an automated way to configure a FIPS-compliant truststore by converting the JVM's default truststore to BCFKS format. The project source is available in `distribution/tools/fips-demo-installer-cli`.

This tool is designed for demo and development purposes. Before deploying to production, carefully review all generated configuration and replace demo settings with production-appropriate values.
{: .warning}

### Prerequisites

Before running the FIPS demo installer, ensure that the following prerequisites are met:

- OpenSearch is installed and the installation directory is accessible.
- You have write permissions to the OpenSearch configuration directory.
- The `jvm.options` file exists in the configuration directory.

### Available commands

The FIPS demo installer provides the following commands.

| Command | Description |
|---------|-------------|
| `generated` | Generates a new BCFKS truststore from the JVM default truststore. |
| `system` | Uses the existing system PKCS11 truststore. |

> | `show-providers` | Shows available security providers and exits (does not change configuration). |

Reviewer: ```suggestion
| `show-providers` | Shows available security providers and exits (does not change the configuration). |
```

### Non-interactive mode

To run the installer in non-interactive mode for automated deployments, use the `-n` or `--non-interactive` flag:

```bash
./bin/opensearch-fips-demo-installer -n
```
{% include copy.html %}

The non-interactive mode runs without prompts and automatically performs the following actions:

- Defaults to generating a new BCFKS truststore.
- Autoconfirms all prompts.
- Generates a secure 24-character password (or uses the one specified using `-p`).
- Selects the first available PKCS11 provider when using the `system` command.

Non-interactive mode is ideal for automated provisioning scripts and configuration management tools.
{: .note}

### Examples

The following are some common command examples for the FIPS demo installer.

On Windows, use `opensearch-fips-demo-installer.bat` instead of the bash script.
{: .note}

Interactive mode (prompts for all choices):
```bash
./bin/opensearch-fips-demo-installer
```
{% include copy.html %}

Non-interactive mode with auto-generated password - overrides existing FIPS configuration:
```bash
./bin/opensearch-fips-demo-installer -n -f
```
{% include copy.html %}

> Generate BCFKS trust store with custom password:

Reviewer: ```suggestion
Generate a BCFKS truststore with a custom password:
```

## Start OpenSearch FIPS Demo Configuration
## WARNING: revise all the lines below before you go into production
################################################################

-Djavax.net.ssl.trustStore=/path/to/opensearch/config/opensearch-fips-truststore.bcfks
-Djavax.net.ssl.trustStorePassword=<your-password>
-Djavax.net.ssl.trustStoreType=BCFKS
-Djavax.net.ssl.trustStoreProvider=BCFIPS
################################################################
```

These properties configure the JVM to use the FIPS-compliant truststore for all SSL/TLS connections if no other truststore is defined.

## Troubleshooting FIPS

This section covers common issues when running OpenSearch in FIPS mode.

### Trust store type not specified

The following error indicates that the FIPS truststore configuration is incomplete or missing from `jvm.options`:

```
Trust store type must be specified using the '-Djavax.net.ssl.trustStoreType' JVM option. Accepted values are PKCS11 and BCFKS.
```

To resolve this issue:

- Verify that you have run the FIPS demo installer successfully.
- Check that `jvm.options` contains the FIPS truststore configuration block.

### Trust store file not found

> If you see an error indicating that the trust store file cannot be found, verify that:

Reviewer: ```suggestion
If you see an error indicating that the truststore file cannot be found, verify that:
```

- The truststore file exists at the specified location.
- OpenSearch has read permissions for the truststore file.

### Certificate conversion failures

Some certificates in the JVM default truststore may not be compatible with the BCFKS format. The installer reports the number of certificates that were successfully converted. Review the output to ensure that critical certificates were converted successfully.

### Keystore password too weak for FIPS mode

When the [OpenSearch keystore]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/security-settings/#password-protection) `$OPENSEARCH_HOME/config/opensearch.keystore` contains a password that does not meet FIPS requirements, OpenSearch fails to start with the following error:

```
org.bouncycastle.crypto.fips.FipsUnapprovedOperationError: password must be at least 112 bits
```

In FIPS mode, Bouncy Castle enforces a minimum password strength of 112 bits, which is approximately 14 characters.

Because the FIPS mode is already active, the `opensearch-keystore passwd` command does not accept the existing weak password. The workaround is to recreate the keystore as follows:

> List existing secrets for backup (if needed):

Reviewer: ```suggestion
1. List existing secrets for backup (if needed):
```

```bash
./bin/opensearch-keystore list
```
{% include copy.html %}

Create a new keystore with a FIPS-compliant password (at least 14 characters):
```bash
./bin/opensearch-keystore create --password
```
{% include copy.html %}

> Re-add any secrets that were stored in the old keystore (if needed):

Reviewer: ```suggestion
1. Re-add any secrets that were stored in the old keystore (if needed):
```

```bash
./bin/opensearch-keystore add <setting-name>
```
{% include copy.html %}

Ensure that your new password is at least 14 characters long and includes a mix of uppercase, lowercase, numbers, and special characters. For security best practices, consider using a password manager to generate and store complex passwords.
{: .note}

## Next steps

After configuring FIPS mode for OpenSearch:

> - Review the [Security configuration]({{site.url}}{{site.baseurl}}/security/configuration/index/) guide for additional security settings.

Reviewer: ```suggestion
- Review the [security configuration]({{site.url}}{{site.baseurl}}/security/configuration/index/) guide for additional security settings.
```
